### PR TITLE
Fix issue 23948 - __FILE__ and __MODULE__ cannot be implicitly converted to const(char)* as default paramenter

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: compiler/src/dmd/expression.d [hunk 1]
```diff
@@ -7101,7 +7101,7 @@ extern (C++) final class FileInitExp : DefaultInitExp
 
         Expression e = new StringExp(loc, s.toDString());
         e = e.expressionSemantic(sc);
-        e = e.castTo(sc, type);
+        e.type = Type.tstring;
```
Comment (maxhaton) [inline compiler/src/dmd/expression.d:7104]: Type should probably be set in the constructor rather than here.

Comment (RazvanN7) [reply compiler/src/dmd/expression.d:7104]: Calling expression semantic and then manually setting the type is pointless. I think that simply calling expression semantic should fix the issue. The cast seems useless.

I see that you simply followed the existing logic, but that seems broken to me.

Comment (RazvanN7) [reply compiler/src/dmd/expression.d:7104]: Note that this code works without this patch:

```d
void foo(const(char)* mod)() {
}

void main() {
    enum string t = __MODULE__;
    foo!t();
}
```

**Comment by ErnyTech** [comment]
I removed the casts and type setting

**Review by RazvanN7 — APPROVED** [review]
Yep, this is fine. Thanks! As suspected, we were doing extra work for nothing.

**Pull request merged**